In [1]:
import pandas as pd

### Importing Drug Review Dataset
Source: https://archive.ics.uci.edu/ml/datasets/Drug+Review+Dataset+%28Drugs.com%29

In [2]:
drugs = pd.read_csv('../Data/Raw_Data/DrugReview.csv')

In [3]:
drugs.shape

(161297, 7)

In [4]:
drugs.head(10)

Unnamed: 0                            drugName  \
0      206461                           Valsartan   
1       95260                          Guanfacine   
2       92703                              Lybrel   
3      138000                          Ortho Evra   
4       35696            Buprenorphine / naloxone   
5      155963                              Cialis   
6      165907                      Levonorgestrel   
7      102654                        Aripiprazole   
8       74811                              Keppra   
9       48928  Ethinyl estradiol / levonorgestrel   

                      condition  \
0  Left Ventricular Dysfunction   
1                          ADHD   
2                 Birth Control   
3                 Birth Control   
4             Opiate Dependence   
5  Benign Prostatic Hyperplasia   
6       Emergency Contraception   
7               Bipolar Disorde   
8                      Epilepsy   
9                 Birth Control   

                                              review  rating  \
0  "It has no side effect, I take it in combinati...     9.0   
1  "My son is halfway through his fourth week of ...     8.0   
2  "I used to take another oral contraceptive, wh...     5.0   
3  "This is my first time using any form of birth...     8.0   
4  "Suboxone has completely turned my life around...     9.0   
5  "2nd day on 5mg started to work with rock hard...     2.0   
6  "He pulled out, but he cummed a bit in me. I t...     1.0   
7  "Abilify changed my life. There is hope. I was...    10.0   
8  " I Ve had  nothing but problems with the Kepp...     1.0   
9  "I had been on the pill for many years. When m...     8.0   

                date  usefulCount  
0       May 20, 2012           27  
1     April 27, 2010          192  
2  December 14, 2009           17  
3   November 3, 2015           10  
4  November 27, 2016           37  
5  November 28, 2015           43  
6      March 7, 2017            5  
7     March 14, 2015           32  
8     August 9, 2016           11  
9   December 8, 2016            1

In [5]:
drugs.isnull().sum()

Unnamed: 0       0
drugName         0
condition      899
review           0
rating           0
date             0
usefulCount      0
dtype: int64

In [6]:
drugs.dropna(inplace = True)

In [7]:
drugs.isnull().sum()

Unnamed: 0     0
drugName       0
condition      0
review         0
rating         0
date           0
usefulCount    0
dtype: int64

In [8]:
drugs.shape

(160398, 7)

In [9]:
drugs['condition'].value_counts()

Birth Control                                              28788
Depression                                                  9069
Pain                                                        6145
Anxiety                                                     5904
Acne                                                        5588
                                                           ...  
Dissociative Identity Disorde                                  1
Hydrocephalus                                                  1
Hyperlipoproteinemia Type III, Elevated beta-VLDL   IDL        1
Q Feve                                                         1
Neutropenia                                                    1
Name: condition, Length: 884, dtype: int64

### Psychiatric Conditions & Medications

Since my background is in research psychology & neuroscience, I decided to focus on the conditions & medications speicifically related to mental health:

In [10]:
psychdrugs = drugs.loc[drugs['condition'].isin(['Depression', 'Anxiety', 'Bipolar Disorde', 'ADHD', 'Anxiety and Stress', 'Major Depressive Disorde', 'Panic Disorde', 'Generalized Anxiety Disorde', 'Schizophrenia', 'Obsessive Compulsive Disorde', 'Schizoaffective Disorde', 'Social Anxiety Disorde', 'Post Traumatic Stress Disorde', 'mance Anxiety',
                    'Borderline Personality Disorde', 'Postpartum Depression', 'Psychosis', 'Sexual Dysfunction, SSRI Induced',
                   'Paranoid Disorde', 'Nightmares'])]

In [11]:
psychdrugs.head()

Unnamed: 0        drugName        condition  \
1        95260      Guanfacine             ADHD   
7       102654    Aripiprazole  Bipolar Disorde   
11       75612  L-methylfolate       Depression   
18      212077     Lamotrigine  Bipolar Disorde   
27       78563        Daytrana             ADHD   

                                               review  rating  \
1   "My son is halfway through his fourth week of ...     8.0   
7   "Abilify changed my life. There is hope. I was...    10.0   
11  "I have taken anti-depressants for years, with...    10.0   
18  "I&#039;ve been on every medicine under the su...    10.0   
27  "Hi all, My son who is 12 was diagnosed when h...    10.0   

                date  usefulCount  
1     April 27, 2010          192  
7     March 14, 2015           32  
11     March 9, 2017           54  
18  November 9, 2014           18  
27  January 12, 2017           11

In [12]:
psychdrugs.shape

(31369, 7)

In [13]:
psychdrugs['condition'].value_counts()

Depression                          9069
Anxiety                             5904
Bipolar Disorde                     4224
ADHD                                3383
Anxiety and Stress                  1663
Major Depressive Disorde            1607
Panic Disorde                       1463
Generalized Anxiety Disorde         1164
Schizophrenia                        594
Obsessive Compulsive Disorde         579
Schizoaffective Disorde              396
Social Anxiety Disorde               389
Post Traumatic Stress Disorde        314
mance Anxiety                        187
Borderline Personality Disorde       151
Postpartum Depression                 84
Psychosis                             73
Sexual Dysfunction, SSRI Induced      69
Paranoid Disorde                      38
Nightmares                            18
Name: condition, dtype: int64

In [14]:
psychdrugs_csv = psychdrugs.to_csv('../Data/Processed_Data/psychdrugs.csv', index = False)

In [15]:
psych = pd.read_csv('../Data/Processed_Data/psychdrugs.csv')

In [16]:
psych.head()

Unnamed: 0        drugName        condition  \
0       95260      Guanfacine             ADHD   
1      102654    Aripiprazole  Bipolar Disorde   
2       75612  L-methylfolate       Depression   
3      212077     Lamotrigine  Bipolar Disorde   
4       78563        Daytrana             ADHD   

                                              review  rating  \
0  "My son is halfway through his fourth week of ...     8.0   
1  "Abilify changed my life. There is hope. I was...    10.0   
2  "I have taken anti-depressants for years, with...    10.0   
3  "I&#039;ve been on every medicine under the su...    10.0   
4  "Hi all, My son who is 12 was diagnosed when h...    10.0   

               date  usefulCount  
0    April 27, 2010          192  
1    March 14, 2015           32  
2     March 9, 2017           54  
3  November 9, 2014           18  
4  January 12, 2017           11

In [17]:
psychdrugs['drugName'].nunique()

259

In [18]:
psychdrugs['condition'].nunique()

20

In [19]:
psychdrugs.head()

Unnamed: 0        drugName        condition  \
1        95260      Guanfacine             ADHD   
7       102654    Aripiprazole  Bipolar Disorde   
11       75612  L-methylfolate       Depression   
18      212077     Lamotrigine  Bipolar Disorde   
27       78563        Daytrana             ADHD   

                                               review  rating  \
1   "My son is halfway through his fourth week of ...     8.0   
7   "Abilify changed my life. There is hope. I was...    10.0   
11  "I have taken anti-depressants for years, with...    10.0   
18  "I&#039;ve been on every medicine under the su...    10.0   
27  "Hi all, My son who is 12 was diagnosed when h...    10.0   

                date  usefulCount  
1     April 27, 2010          192  
7     March 14, 2015           32  
11     March 9, 2017           54  
18  November 9, 2014           18  
27  January 12, 2017           11

### Adding Drug Classes

Decided it would be interesting to also look at the drug class of each of these 259 drugs; I looked this up by-hand (source: https://medlineplus.gov/), added to Google Sheets & exported as a csv. Will merge below:

In [20]:
classes = pd.read_csv('../Data/Raw_Data/DrugClasses.csv')

In [21]:
classes.head()

Drug Name Drug Class
0    Sertraline       SSRI
1  Escitalopram       SSRI
2    Citalopram       SSRI
3       Lexapro       SSRI
4   Venlafaxine       SNRI

In [22]:
classes.shape

(259, 2)

In [23]:
classes = classes.rename(columns = {'Drug Name': 'drugName'})

In [24]:
psych.shape

(31369, 7)

In [25]:
classes.dtypes

drugName      object
Drug Class    object
dtype: object

In [26]:
psych.dtypes

Unnamed: 0       int64
drugName        object
condition       object
review          object
rating         float64
date            object
usefulCount      int64
dtype: object

In [27]:
merged = pd.merge(psych, classes, on = 'drugName', how = 'left')

In [28]:
merged.head()

Unnamed: 0        drugName        condition  \
0       95260      Guanfacine             ADHD   
1      102654    Aripiprazole  Bipolar Disorde   
2       75612  L-methylfolate       Depression   
3      212077     Lamotrigine  Bipolar Disorde   
4       78563        Daytrana             ADHD   

                                              review  rating  \
0  "My son is halfway through his fourth week of ...     8.0   
1  "Abilify changed my life. There is hope. I was...    10.0   
2  "I have taken anti-depressants for years, with...    10.0   
3  "I&#039;ve been on every medicine under the su...    10.0   
4  "Hi all, My son who is 12 was diagnosed when h...    10.0   

               date  usefulCount                         Drug Class  
0    April 27, 2010          192                     alpha agonists  
1    March 14, 2015           32            atypical antipsychotics  
2     March 9, 2017           54                          B vitamin  
3  November 9, 2014           18                    anticonvulsants  
4  January 12, 2017           11  central nervous system stimulants

In [29]:
merged.shape

(31369, 8)

In [30]:
merged['Drug Class'].value_counts()

SSRI                                  8975
atypical antipsychotics               4506
SNRI                                  4333
central nervous system stimulants     2722
benzodiazepines                       2669
anticonvulsants                       2181
NRDI                                  1394
SSRI/5HT-1A Partial Agonist            782
TeCA                                   524
anxiolytics                            466
antihistamines                         460
tricyclic antidepressants              342
beta blockers                          321
Anxiolytics                            261
MAOI                                   215
alpha agonists                         214
antimanic agents                       156
B vitamin                              117
SARI                                   114
antihypertensives                      113
conventional antipsychotics            103
opiate analgesics                       86
wakefulness promoting agents            72
MAO-B      

Decided to drop Drug Classes with <100 values; classes will still be imbalanced but <100 seem like they could distort the data/be outliers. Will also drop 'Unnamed:0', 'date' columns as I will definitely not be using in analysis.

In [31]:
merged = merged.rename(columns = {'Drug Class': 'Drug_Class', 'drugName' : 'Drug_Name', 'condition' : 'Condition',
                                 'review': 'Review', 'rating': 'Rating'})

In [32]:
drop_values = ['opiate analgesics', 'wakefulness promoting agents',
              'MAO-B', 'psychotherapeutic combos', 'central nervous system depressants', 'serotonin modulators',
               'tranquilizers', 'phosphodiesterase inhibitors','tetracyclic antidepressants',
               'aminoketone', 'SAMe', 'alpha-2 antagonists', 'cannabinoids', 'benzothiazoles',
               'antidyskinetic', 'herbals', 'amino acid', 'antilipemic agents', 'bipolar disorder agents',
              'dopamine reuptake inhibitors', 'phosphodiesterase (PDE) inhibitors', 'serotonin modulators '] 

merged = merged[merged.Drug_Class.isin(drop_values) == False]        

In [33]:
merged['Drug_Class'].nunique()

21

In [34]:
merged = merged.drop(columns = 'Unnamed: 0', axis = 1)

In [35]:
merged = merged.drop(columns = 'date', axis = 1)

In [36]:
merged.dtypes

Drug_Name       object
Condition       object
Review          object
Rating         float64
usefulCount      int64
Drug_Class      object
dtype: object

In [37]:
merged.isnull().sum()

Drug_Name      0
Condition      0
Review         0
Rating         0
usefulCount    0
Drug_Class     0
dtype: int64

In [38]:
Merged_data = merged.to_csv('../Data/Processed_Data/Merged.csv', index = False)

### Cleaning Drug Review Text

In [44]:
df = pd.read_csv('../Data/Processed_Data/Merged.csv')

In [45]:
df.head(20)

Drug_Name                 Condition  \
0       Guanfacine                      ADHD   
1     Aripiprazole           Bipolar Disorde   
2   L-methylfolate                Depression   
3      Lamotrigine           Bipolar Disorde   
4         Daytrana                      ADHD   
5           Ativan             Panic Disorde   
6       Sertraline                Depression   
7         Trilafon                 Psychosis   
8       Clonazepam             Panic Disorde   
9      Venlafaxine                Depression   
10         Intuniv                      ADHD   
11    Aripiprazole             Schizophrenia   
12          Latuda           Bipolar Disorde   
13      Effexor XR                   Anxiety   
14      Effexor XR                Depression   
15      Sertraline             Panic Disorde   
16      Wellbutrin                Depression   
17         Lithium           Bipolar Disorde   
18      Fluoxetine  Major Depressive Disorde   
19    Aripiprazole                Depression   

                                               Review  Rating  usefulCount  \
0   "My son is halfway through his fourth week of ...     8.0          192   
1   "Abilify changed my life. There is hope. I was...    10.0           32   
2   "I have taken anti-depressants for years, with...    10.0           54   
3   "I&#039;ve been on every medicine under the su...    10.0           18   
4   "Hi all, My son who is 12 was diagnosed when h...    10.0           11   
5   "Honestly, I have been taking ativan for 2 yea...     6.0           47   
6   "1 week on Zoloft for anxiety and mood swings....     8.0            3   
7   "I had a similar experience.  Tremors in hands...     9.0           45   
8   "This medication changed my life.  My panic at...     9.0           30   
9   "my gp started me on Venlafaxine yesterday to ...     4.0            3   
10  "Intuniv did not work for my son; he was bounc...     1.0           23   
11  "Abilify 20 mg.\r\nI am a patient diagnosed wi...    10.0           26   
12  "I have had great experience so far with Latud...     8.0           39   
13  "Was on this med for 5 years. Worked fine but ...     6.0           23   
14  "This medicine saved my life. I was at my wits...    10.0          166   
15  "These reviews helped me so much when I first ...    10.0           28   
16  "Started taking it and I slept well at night a...     4.0           15   
17  "I appreciate Lithium. Although my thyroid lev...    10.0           27   
18  "I started Prozac as one of my first anti depr...     2.0           18   
19  "Intake Effexor XR 375 mg, and lorazepam for d...     4.0           33   

                           Drug_Class  
0                      alpha agonists  
1             atypical antipsychotics  
2                           B vitamin  
3                     anticonvulsants  
4   central nervous system stimulants  
5                     benzodiazepines  
6                                SSRI  
7         conventional antipsychotics  
8                     benzodiazepines  
9                                SNRI  
10                     alpha agonists  
11            atypical antipsychotics  
12            atypical antipsychotics  
13                               SNRI  
14                               SNRI  
15                               SSRI  
16                               NRDI  
17                   antimanic agents  
18                               SSRI  
19            atypical antipsychotics

In [46]:
df.isnull().sum()

Drug_Name      0
Condition      0
Review         0
Rating         0
usefulCount    0
Drug_Class     0
dtype: int64

In [50]:
import re

In [56]:
#Remove quotations 
df['Review'] = df['Review'].str.replace('"', '', regex = True)
    
#Remove punctuation
df['Review'] = df['Review'].str.replace('[^\w\s]', '', regex = True)
    
#Remove formatting for apostrophes
df['Review'] = df['Review'].str.replace('&#039;', '', regex = True)
    
#Transform to lowercase
df['Review'] = df['Review'].str.lower()
    
#Remove 'mg'
df['Review'] = df['Review'].str.replace('mg', '', regex = True)
    
#Remove periods and commas
df['Review'] = df['Review'].str.replace('.','', regex = True).str.replace(',', '', regex = True)

In [57]:
df.head()

Drug_Name        Condition  \
0      Guanfacine             ADHD   
1    Aripiprazole  Bipolar Disorde   
2  L-methylfolate       Depression   
3     Lamotrigine  Bipolar Disorde   
4        Daytrana             ADHD   

                                              Review  Rating  usefulCount  \
0  my son is halfway through his fourth week of i...     8.0          192   
1  abilify changed my life there is hope i was on...    10.0           32   
2  i have taken antidepressants for years with so...    10.0           54   
3  i039ve been on every medicine under the sun it...    10.0           18   
4  hi all my son who is 12 was diagnosed when he ...    10.0           11   

                          Drug_Class  \
0                     alpha agonists   
1            atypical antipsychotics   
2                          B vitamin   
3                    anticonvulsants   
4  central nervous system stimulants   

                                             Reviews  
0  "My son is halfway through his fourth week of ...  
1  "Abilify changed my life There is hope I was o...  
2  "I have taken anti-depressants for years with ...  
3  "I&#039;ve been on every medicine under the su...  
4  "Hi all My son who is 12 was diagnosed when he...

In [91]:
df = df.drop(columns = 'Reviews', axis = 1)

In [92]:
df.dtypes

Drug_Name       object
Condition       object
Review          object
Rating         float64
usefulCount      int64
Drug_Class      object
dtype: object

In [99]:
#Remove numbers
def remove_num(col):
    col = re.sub('[^A-za-z]', ' ', col)
    return col.lower()

In [107]:
df['Reviews'] = df['Review'].apply(remove_num)
df.head()

Drug_Name        Condition  \
0      Guanfacine             ADHD   
1    Aripiprazole  Bipolar Disorde   
2  L-methylfolate       Depression   
3     Lamotrigine  Bipolar Disorde   
4        Daytrana             ADHD   

                                              Review  Rating  usefulCount  \
0  my son is halfway through his fourth week of i...     8.0          192   
1  abilify changed my life there is hope i was on...    10.0           32   
2  i have taken antidepressants for years with so...    10.0           54   
3  ive been on every medicine under the sun it se...    10.0           18   
4  hi all my son who is  was diagnosed when he wa...    10.0           11   

                          Drug_Class  \
0                     alpha agonists   
1            atypical antipsychotics   
2                          B vitamin   
3                    anticonvulsants   
4  central nervous system stimulants   

                                             Reviews  
0  my son is halfway through his fourth week of i...  
1  abilify changed my life there is hope i was on...  
2  i have taken antidepressants for years with so...  
3  ive been on every medicine under the sun it se...  
4  hi all my son who is  was diagnosed when he wa...

In [105]:
#Remove line spacing
df['Reviews'] = df['Reviews'].str.replace('\r\n', '', regex = True)

In [111]:
#Checking different reviews to ensure cleaning worked:
display(df['Reviews'].iloc[1000])

'excellent results effexor became noneffective after many years of use and another life impacting event was recently severely depressed and suicidal for about  years found cymbalta  years ago and it changed my life for the better i think this medication is fantastic i have been diagnosed with major depressive disorder and general anxiety disorder dual diagnosed this condition has been with me for many years runs in the family ie mother was severely depressed cymbalta  per day snapped me out of it under doctors orders i discontinued current ssri at the time and started right up with cymbalta day within  days i was out of my depressed state it has to be the norepinephrine aspect that made the difference'

In [112]:
#Dropping duplicage Review column
df = df.drop(columns = 'Review', axis = 1)

### Exporting Cleaned Data to csv

In [113]:
df_csv = df.to_csv('../Data/Processed_Data/cleaned.csv')